File: IsometricRPG/IsometricRPG/World/WorldManager.swift
import SpriteKit

/// Manages the infinite, chunk-based world. Loads/unloads chunks around the player.
final class WorldManager {
    private(set) var loadedChunks: [ChunkCoord: Chunk] = [:]
    let rootNode = SKNode()
    private let noise: PerlinNoise
    private let moistureNoise: PerlinNoise
    private var lastPlayerChunk: ChunkCoord?

    init() {
        noise = PerlinNoise(seed: Constants.worldSeed)
        moistureNoise = PerlinNoise(seed: Constants.worldSeed &+ 12345)
        rootNode.name = "world"
    }

    // MARK: - Chunk Management

    /// Call every frame with the player's world position.
    /// Returns (loaded, unloaded) chunk coords for this tick.
    @discardableResult
    func updateAroundPlayer(worldPosition: CGPoint) -> (loaded: [ChunkCoord], unloaded: [ChunkCoord]) {
        let playerChunk = ChunkCoord.containing(worldPosition: worldPosition)

        // Skip if player hasn't moved chunks
        if playerChunk == lastPlayerChunk { return ([], []) }
        lastPlayerChunk = playerChunk

        var loaded: [ChunkCoord] = []
        var unloaded: [ChunkCoord] = []

        // Load chunks within radius
        let r = Constants.loadRadius
        for dy in -r...r {
            for dx in -r...r {
                let coord = ChunkCoord(x: playerChunk.x + dx, y: playerChunk.y + dy)
                if loadedChunks[coord] == nil {
                    let chunk = Chunk(coord: coord, noise: noise, moistureNoise: moistureNoise)
                    loadedChunks[coord] = chunk
                    rootNode.addChild(chunk.node)
                    loaded.append(coord)
                }
            }
        }

        // Unload distant chunks
        let toUnload = loadedChunks.keys.filter { coord in
            coord.chebyshevDistance(to: playerChunk) > Constants.unloadRadius
        }
        for coord in toUnload {
            if let chunk = loadedChunks.removeValue(forKey: coord) {
                chunk.node.removeFromParent()
                unloaded.append(coord)
            }
        }

        return (loaded, unloaded)
    }

    /// Force-load chunks around a position (used at game start).
    func initialLoad(around worldPosition: CGPoint) {
        updateAroundPlayer(worldPosition: worldPosition)
    }

    // MARK: - Tile Queries

    func tileAt(worldCol: Int, worldRow: Int) -> TileType? {
        let chunkCoord = ChunkCoord.containing(worldCol: worldCol, worldRow: worldRow)
        guard let chunk = loadedChunks[chunkCoord] else { return nil }
        let localCol = worldCol - chunkCoord.worldOriginCol
        let localRow = worldRow - chunkCoord.worldOriginRow
        return chunk.tileAt(localCol: localCol, localRow: localRow)
    }

    func elevationAt(worldCol: Int, worldRow: Int) -> Int? {
        let chunkCoord = ChunkCoord.containing(worldCol: worldCol, worldRow: worldRow)
        guard let chunk = loadedChunks[chunkCoord] else { return nil }
        let localCol = worldCol - chunkCoord.worldOriginCol
        let localRow = worldRow - chunkCoord.worldOriginRow
        return chunk.elevationAt(localCol: localCol, localRow: localRow)
    }

    func elevationAt(worldPosition: CGPoint) -> Int? {
        let col = Int(floor(worldPosition.x + 0.5))
        let row = Int(floor(worldPosition.y + 0.5))
        return elevationAt(worldCol: col, worldRow: row)
    }

    func isWalkable(worldCol: Int, worldRow: Int) -> Bool {
        return tileAt(worldCol: worldCol, worldRow: worldRow)?.isWalkable ?? false
    }

    func isWalkable(worldPosition: CGPoint) -> Bool {
        let col = Int(floor(worldPosition.x + 0.5))
        let row = Int(floor(worldPosition.y + 0.5))
        return isWalkable(worldCol: col, worldRow: row)
    }

    /// Check if movement from one position to another is walkable (considering elevation).
    func isWalkable(from: CGPoint, to: CGPoint) -> Bool {
        let fromCol = Int(floor(from.x + 0.5))
        let fromRow = Int(floor(from.y + 0.5))
        let toCol = Int(floor(to.x + 0.5))
        let toRow = Int(floor(to.y + 0.5))

        // Check base walkability of destination
        guard let toTile = tileAt(worldCol: toCol, worldRow: toRow),
              toTile.isWalkable else {
            return false
        }

        // Check elevation difference
        guard let fromElev = elevationAt(worldCol: fromCol, worldRow: fromRow),
              let toElev = elevationAt(worldCol: toCol, worldRow: toRow) else {
            return false
        }

        let heightDiff = abs(toElev - fromElev)
        if heightDiff > Constants.maxWalkableElevationDiff {
            return false
        }

        return true
    }

    func biomeAt(worldPosition: CGPoint) -> Biome? {
        let chunkCoord = ChunkCoord.containing(worldPosition: worldPosition)
        return loadedChunks[chunkCoord]?.biome
    }

    func chunkAt(worldPosition: CGPoint) -> Chunk? {
        let coord = ChunkCoord.containing(worldPosition: worldPosition)
        return loadedChunks[coord]
    }

    /// Get newly loaded chunks that haven't had enemies spawned yet.
    func chunksNeedingEnemySpawn() -> [Chunk] {
        return loadedChunks.values.filter { !$0.spawnedEnemies }
    }

    /// Get newly loaded chunks that haven't had items spawned yet.
    func chunksNeedingItemSpawn() -> [Chunk] {
        return loadedChunks.values.filter { !$0.spawnedItems }
    }
}

File: IsometricRPGMono/World/WorldManager.cs
using Microsoft.Xna.Framework;

namespace IsometricRPG;

public class WorldManager
{
    private readonly Dictionary<ChunkCoord, Chunk> _chunks = new();
    private readonly PerlinNoise _elevNoise;
    private readonly PerlinNoise _moistureNoise;
    private ChunkCoord _lastPlayerChunk;
    private bool _firstUpdate = true;

    public IReadOnlyDictionary<ChunkCoord, Chunk> LoadedChunks => _chunks;

    public WorldManager()
    {
        _elevNoise    = new PerlinNoise(Constants.WorldSeed);
        _moistureNoise = new PerlinNoise(Constants.WorldSeed + 12345);
    }

    /// Call every frame. Returns (loaded, unloaded) coord lists.
    public (List<ChunkCoord> loaded, List<ChunkCoord> unloaded) UpdateAroundPlayer(Vector2 worldPos)
    {
        int tileX = (int)MathF.Floor(worldPos.X);
        int tileY = (int)MathF.Floor(worldPos.Y);
        var playerChunk = ChunkCoord.FromWorld(tileX, tileY);

        if (!_firstUpdate && playerChunk == _lastPlayerChunk)
            return (new List<ChunkCoord>(), new List<ChunkCoord>());

        _firstUpdate = false;
        _lastPlayerChunk = playerChunk;

        var loaded   = new List<ChunkCoord>();
        var unloaded = new List<ChunkCoord>();

        int r = Constants.LoadRadius;
        for (int dy = -r; dy <= r; dy++)
            for (int dx = -r; dx <= r; dx++)
            {
                var coord = new ChunkCoord(playerChunk.X + dx, playerChunk.Y + dy);
                if (!_chunks.ContainsKey(coord))
                {
                    _chunks[coord] = new Chunk(coord, _elevNoise, _moistureNoise);
                    loaded.Add(coord);
                }
            }

        var toUnload = _chunks.Keys
            .Where(c => c.ChebyshevDistance(playerChunk) > Constants.UnloadRadius)
            .ToList();
        foreach (var c in toUnload)
        {
            _chunks.Remove(c);
            unloaded.Add(c);
        }

        return (loaded, unloaded);
    }

    public void InitialLoad(Vector2 worldPos) => UpdateAroundPlayer(worldPos);

    // ---- Tile queries ----

    public TileType? TileAt(int worldCol, int worldRow)
    {
        var coord = ChunkCoord.FromWorld(worldCol, worldRow);
        if (!_chunks.TryGetValue(coord, out var chunk)) return null;
        var (ox, oy) = coord.WorldOrigin();
        return chunk.TileAt(worldCol - ox, worldRow - oy);
    }

    public float? ElevationAt(int worldCol, int worldRow)
    {
        var coord = ChunkCoord.FromWorld(worldCol, worldRow);
        if (!_chunks.TryGetValue(coord, out var chunk)) return null;
        var (ox, oy) = coord.WorldOrigin();
        return chunk.ElevationAt(worldCol - ox, worldRow - oy);
    }

    public bool IsWalkable(Vector2 worldPos)
    {
        int col = (int)MathF.Floor(worldPos.X + 0.5f);
        int row = (int)MathF.Floor(worldPos.Y + 0.5f);
        return TileAt(col, row)?.IsWalkable() ?? false;
    }

    public bool IsWalkable(Vector2 from, Vector2 to)
    {
        if (!IsWalkable(to)) return false;
        int fc = (int)MathF.Floor(from.X + 0.5f); int fr = (int)MathF.Floor(from.Y + 0.5f);
        int tc = (int)MathF.Floor(to.X   + 0.5f); int tr = (int)MathF.Floor(to.Y   + 0.5f);
        float? fe = ElevationAt(fc, fr);
        float? te = ElevationAt(tc, tr);
        if (fe == null || te == null) return false;
        // Elevation is [0,1]; multiply by 100 to compare against integer-percent threshold.
        return MathF.Abs(te.Value - fe.Value) * 100f <= Constants.MaxWalkableElevationDiff;
    }

    public Biome? BiomeAt(Vector2 worldPos)
    {
        var coord = ChunkCoord.FromWorld((int)MathF.Floor(worldPos.X), (int)MathF.Floor(worldPos.Y));
        return _chunks.TryGetValue(coord, out var c) ? c.Biome : null;
    }

    public Chunk? ChunkAt(Vector2 worldPos)
    {
        var coord = ChunkCoord.FromWorld((int)MathF.Floor(worldPos.X), (int)MathF.Floor(worldPos.Y));
        return _chunks.TryGetValue(coord, out var c) ? c : null;
    }

    public bool IsPositionLoaded(int worldCol, int worldRow)
        => _chunks.ContainsKey(ChunkCoord.FromWorld(worldCol, worldRow));

    public List<Chunk> ChunksNeedingItemSpawn()  => _chunks.Values.Where(c => !c.SpawnedItems).ToList();
    public List<Chunk> ChunksNeedingEnemySpawn() => _chunks.Values.Where(c => !c.SpawnedEnemies).ToList();
}
